samples/pipeline/unit-of-work/Core_6/Endpoint/Program.cs

```diff
@@ -0,0 +1,142 @@
+using System;
+using System.Threading.Tasks;
+using NServiceBus;
+using NServiceBus.Pipeline;
+
+class Program
+{
+    static void Main()
+    {
+        AsyncMain().GetAwaiter().GetResult();
+    }
+
+    static async Task AsyncMain()
+    {
+        Console.Title = "Samples.UnitOfWork.Endpoint";
+        var endpointConfiguration = new EndpointConfiguration("Samples.UnitOfWork.Endpoint");
+        endpointConfiguration.UsePersistence<InMemoryPersistence>();
+        endpointConfiguration.EnableInstallers();
+        endpointConfiguration.SendFailedMessagesTo("error");
+
+        endpointConfiguration.Pipeline.Register(typeof(MyUowBehavior), "Manages the session");
+        endpointConfiguration.RegisterComponents(r => r.ConfigureComponent(b => new IMySession(), DependencyLifecycle.InstancePerUnitOfWork));
+
+        var endpointInstance = await Endpoint.Start(endpointConfiguration)
+            .ConfigureAwait(false);
+
+        var key = default(ConsoleKeyInfo);
+
+        Console.WriteLine("Press any key to send messages, 'q' to exit");
+        while (key.KeyChar != 'q')
+        {
+            key = Console.ReadKey();
+
+            for (var i = 0; i < 4; i++)
+            {
+                var options = new SendOptions();
+
+                options.SetHeader("tennant", "tennant" + i);
+                options.RouteToThisEndpoint();
+                await endpointInstance.Send(new MyMessage(), options);
+            }
+        }
+
+        await endpointInstance.Stop()
+            .ConfigureAwait(false);
+    }
+}
+
+class MyUowBehavior : Behavior<IIncomingPhysicalMessageContext>
+{
+
+    public override async Task Invoke(IIncomingPhysicalMessageContext context, Func<Task> next)
+    {
+
+        var tennant = context.MessageHeaders["tennant"];
+        var session = context.Builder.Build<IMySession>();
```

Comment:
Agree with @DavidBoike 

I'd say the only dependency it should have is a dependency to a session factory which is a singleton. Then it explicitly creates a sessions and adds it to the extension part of the context.